model/roles.go

```diff
@@ -373,38 +385,117 @@ func (m *RoleManifest) LookupRole(roleName string) *Role {
 func (m *RoleManifest) resolveLinks() validation.ErrorList {
 	errors := make(validation.ErrorList, 0)
 
-	// Resolve the consumers
+	// Build mappings of providers by name, and by type.  Note that the names
+	// involved here are the aliases, where appropriate.
+	providersByName := make(map[string]jobProvidesInfo)
+	providersByType := make(map[string][]jobProvidesInfo)
 	for _, role := range m.Roles {
-		// Prefer finding providers in the same role
-		providingRoles := append(Roles{role}, m.Roles...)
-		for _, job := range role.Jobs {
-		iterateOverConsumers:
-			for idx, consumer := range job.LinkConsumes {
-				for _, providingRole := range providingRoles {
-					for _, providingJob := range providingRole.Jobs {
-						for _, provider := range providingJob.LinkProvides {
-							if consumer.Type != "" && consumer.Type != provider.Type {
-								continue
-							}
-							if consumer.Name != "" && consumer.Name != provider.Name {
-								continue
-							}
-							if consumer.Name == "" {
-								consumer.Name = provider.Name
-							}
-							consumer.RoleName = providingRole.Name
-							consumer.JobName = providingJob.Name
-							continue iterateOverConsumers
-						}
-					}
+		for _, roleJob := range role.RoleJobs {
+			var availableProviders []string
+			for availableName, availableProvider := range roleJob.Job.AvailableProviders {
+				availableProviders = append(availableProviders, availableName)
+				providersByType[availableProvider.Type] = append(providersByType[availableProvider.Type], jobProvidesInfo{
+					jobLinkInfo: jobLinkInfo{
+						Name:     availableProvider.Name,
+						Type:     availableProvider.Type,
+						RoleName: role.Name,
+						JobName:  roleJob.Name,
+					},
+					Properties: availableProvider.Properties,
+				})
+			}
+			for name, provider := range roleJob.ExportedProviders {
+				info, ok := roleJob.Job.AvailableProviders[name]
+				if !ok {
+					errors = append(errors, validation.NotFound(
+						fmt.Sprintf("roles[%s].jobs[%s].provides[%s]", role.Name, roleJob.Name, name),
+						fmt.Sprintf("Provider not found; available providers: %v", availableProviders)))
+					continue
+				}
+				if provider.Alias != "" {
+					name = provider.Alias
 				}
+				providersByName[name] = jobProvidesInfo{
+					jobLinkInfo: jobLinkInfo{
+						Name:     info.Name,
+						Type:     info.Type,
+						RoleName: role.Name,
+						JobName:  roleJob.Name,
+					},
+					Properties: info.Properties,
+				}
+			}
+		}
+	}
 
-				// No valid providers
-				if !consumer.Optional {
+	// Resolve the consumers
+	for _, role := range m.Roles {
+		for _, roleJob := range role.RoleJobs {
+			expectedConsumers := make([]jobConsumesInfo, len(roleJob.Job.DesiredConsumers))
+			copy(expectedConsumers, roleJob.Job.DesiredConsumers)
+			if roleJob.ResolvedConsumers == nil {
+				roleJob.ResolvedConsumers = make(map[string]jobConsumesInfo)
```

Comment:
Why create an empty map? It is not getting filled in afaict.

And `range` can iterate over it just fine (it already does so in `LoadRoleManifest`).

Reply:
It does get filled in the `range expectedConsumers` loop later in this function. But there we get the value (probably the zero value, if the consumer wasn't explicitly specified in the role manifest), which mean we should ensure the map isn't nil (but an empty map) before that point.

Yes, `range` can iterate over nil maps/slices. But indexing a nil map panics (whereas indexing an empty map just gives the zero value).

Reply:
What I see in that loop is just

```
					info := roleJob.ResolvedConsumers[name]
					info.Name = provider.Name
```

Since `ResolvedConsumers` is the empty map, `info` will be `nil`, so the assignment to `info.Name` will panic. If it never panics, it means we never get to that spot. Or I still don't understand it.

Reply:
Ah, nm, I missed the reference to the `zero` value, and that we get a value, not a pointer...

Reply:
Yep, if we don't create the map, then `info := roleJob.ResolvedConsumers[name]` will panic (since it involves dereferencing a nil pointer). But once that map exists (but is empty), the result is just an empty struct.

Reply:
I think I would still like to see the initialization either at the place where the whole struct is loaded (`LoadRoleManifest`?), or directly before the place where it _must_ be set to avoid the panic. Right now it is in a spot where it looks like it isn't really useful.